demos/TicTacToe/Classes/tic_tac_toe_layer.cc

```diff
@@ -449,101 +425,155 @@ TicTacToeLayer::TicTacToeLayer(string game_uuid,
-      WaitForCompletion(ref_.Child("game_over").SetValue(true), "setGameOver");
-    } else if (remaining_tiles_.size() == 0) {
-      // Set game_outcome_ to reflect the game ended in a tie.
-      game_outcome_ = kGameTied;
-      WaitForCompletion(ref_.Child("game_over").SetValue(true), "setGameOver");
+      player_index_ = kPlayerTwo;
+      awaiting_opponenet_move_ = true;
     }
   }
+}
+// Creates and sets the listeners as follows.
+// ExpectedValueListeners: total_players & game_over.
+// SampleValueListener: current_player_index & last_move.
+void TicTacToeLayer::InitializeDatabaseListeners() {
+  // total_player_listener_ and CurrentPlayerIndexListener listener is set up
+  // to recognise when the desired players have connected & when turns
+  // alternate.
+  total_player_listener_ = make_unique<ExpectValueListener>(kNumberOfPlayers);
+  game_over_listener_ = make_unique<ExpectValueListener>(true);
 
-  // Shows the end game label for kEndGameFramesMax to show the result of the
-  // game.
-  else if (game_over_listener_->got_value()) {
-    if (game_outcome_ == kGameDisbanded &&
-        remaining_tiles_.size() != kNumberOfTiles) {
-      game_outcome_ = kGameWon;
-    }
-    game_over_label_->setString(kGameOverStrings[game_outcome_]);
-    end_game_frames_++;
-    if (end_game_frames_ > kEndGameFramesMax) {
-      // Remove the game from existence and update the user's record before
-      // swap back scenes.
-      WaitForCompletion(database_->GetReference("game_data")
-                            .Child(join_game_uuid_)
-                            .RemoveValue(),
-                        "removeGameUuid");
-      ref_ = database_->GetReference("users").Child(user_uid_);
-
-      // Updates user record unless the game was disbanded.
-      if (game_outcome_ != kGameDisbanded) {
-        auto future_record =
-            ref_.Child(kGameOutcomeStrings[game_outcome_]).GetValue();
-        WaitForCompletion(future_record, "getPreviousOutcomeRecord");
-        WaitForCompletion(
-            ref_.Child(kGameOutcomeStrings[game_outcome_])
-                .SetValue(future_record.result()->value().int64_value() + 1),
-            "setGameOutcomeRecord");
-      }
+  current_player_index_listener_ = make_unique<SampleValueListener>();
+  last_move_listener_ = make_unique<SampleValueListener>();
 
-      Director::getInstance()->popScene();
+  ref_.Child("total_players").AddValueListener(total_player_listener_.get());
+  ref_.Child("game_over").AddValueListener(game_over_listener_.get());
+  ref_.Child("current_player_index_")
+      .AddValueListener(current_player_index_listener_.get());
+  ref_.Child("last_move").AddValueListener(last_move_listener_.get());
+}
+
+// Counts down the end_game_frames, and updates the user data based on the game
+// outcome.
+void TicTacToeLayer::EndGame() {
+  if (game_outcome_ == kGameDisbanded &&
+      remaining_tiles_.size() != kNumberOfTiles) {
+    game_outcome_ = kGameWon;
+  }
```

Comment:
nit: Add a new line after the }

Reply:
Doesn't show as outdated because new line was below snippet but this is done.